# Multiple Linear Regression Project  

## Tesla Model Y Price Prediction | End-to-End ML Project  

This project focuses on predicting the price of Tesla Model Y cars using **Multiple Linear Regression**.

It covers the full machine learning workflow, including data preprocessing, exploratory data analysis (EDA), feature selection, and model training.

---

## Dataset  

The dataset used in this project is available on Kaggle: [Dataset Link](https://www.kaggle.com/datasets/aiwithcagri/multiple-linear-regression-tesla-model-y-price)
  

---

## Objective  

The main objective is to build a regression model that can accurately estimate car prices based on key features such as:
- year
- km  
- color 

---

## Workflow  

The project follows these steps:

1. **Data Cleaning**

2. **Exploratory Data Analysis (EDA)**

3. **Feature Engineering** 

4. **Model Building**

5. **Model Evaluation**

---

## Tools & Technologies  

- Python  
- Pandas  
- NumPy  
- Scikit-learn  

---

## Key Learnings  

- Understanding how multiple features affect predictions  
- Importance of proper data preprocessing  
- Handling categorical variables correctly  
- Building and interpreting regression models  

---

## Next Steps  

- Apply Polynomial Regression  
- Compare with other regression models (SVR, Decision Tree, Random Forest)  
- Improve model performance with feature engineering  


# **Importing The Libaries**

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# **Data preprocessing**

# **Data Cleaning Check Template**
This template is designed to quickly assess the quality of any dataset before building machine learning models or performing analysis.

It provides a structured overview of the dataset by checking for common data issues such as:

- Missing values

- Duplicate rows

- Incorrect data types

- Outliers

- Distribution of numerical features

- Categorical feature consistency

**What This Template Does**

- Displays basic dataset information (shape, data types)

- Identifies missing values and duplicates

- Summarizes numerical and categorical features

- Detects potential outliers using the IQR method

- Highlights columns with low unique values for quick inspection

How to Use

1. Load your dataset using Pandas  

2. Call the function:


In [ ]:
def data_quality_report(df):

    print("DATA QUALITY REPORT")
    
    # Print a separator line for better readability
    
    print("=" * 50)
    print("BASIC INFO")
    print("=" * 50)
    
    # Show general information about the dataset (columns, data types, non-null values)
    print(df.info())
    
    # Show number of rows and columns
    print("\n" + "=" * 50)
    print("SHAPE OF DATA")
    print("=" * 50)
    print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")
    
    # Check for missing (null) values in each column
    print("\n" + "=" * 50)
    print("MISSING VALUES")
    print("=" * 50)
    missing = df.isnull().sum()
    
    # Display only columns that have missing values
    print(missing[missing > 0])
    
    # Check for duplicate rows
    print("\n" + "=" * 50)
    print("DUPLICATES")
    print("=" * 50)
    print(f"Duplicate rows: {df.duplicated().sum()}")
    
    # Display data types of each column
    print("\n" + "=" * 50)
    print("DATA TYPES")
    print("=" * 50)
    print(df.dtypes)
    
    # Summary statistics for numerical columns (mean, std, min, max, etc.)
    print("\n" + "=" * 50)
    print("NUMERICAL SUMMARY")
    print("=" * 50)
    print(df.describe())
    
    # Summary for categorical (object) columns
    print("\n" + "=" * 50)
    print("CATEGORICAL SUMMARY")
    print("=" * 50)
    print(df.describe(include=['object']))
    
    # Show unique values for columns with low number of distinct values
    # Useful for detecting categories, errors, or inconsistencies
    print("\n" + "=" * 50)
    print("UNIQUE VALUES (LOW CARDINALITY)")
    print("=" * 50)
    for col in df.columns:
        if df[col].nunique() < 10:  # Only show columns with few unique values
            print(f"{col}: {df[col].unique()}")
    
    # Detect outliers using the IQR (Interquartile Range) method
    print("\n" + "=" * 50)
    print("OUTLIERS CHECK (IQR METHOD)")
    print("=" * 50)
    
    # Loop through only numerical columns
    for col in df.select_dtypes(include=np.number).columns:
        Q1 = df[col].quantile(0.25)  # 25th percentile
        Q3 = df[col].quantile(0.75)  # 75th percentile
        IQR = Q3 - Q1  # Interquartile range
        
        # Count rows that fall outside the normal range
        outliers = df[(df[col] < Q1 - 1.5 * IQR) | (df[col] > Q3 + 1.5 * IQR)]
        print(f"{col}: {len(outliers)} outliers")

# **Load dataset**
Apply Data Cleaning Check Template

In [3]:
dataset = pd.read_csv("teslay.csv")
data_quality_report(dataset)

DATA QUALITY REPORT
BASIC INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 255 entries, 0 to 254
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   year    255 non-null    int64  
 1   km      255 non-null    float64
 2   color   255 non-null    object 
 3   price   255 non-null    int64  
dtypes: float64(1), int64(2), object(1)
memory usage: 8.1+ KB
None

SHAPE OF DATA
Rows: 255, Columns: 4

MISSING VALUES
Series([], dtype: int64)

DUPLICATES
Duplicate rows: 4

DATA TYPES
year       int64
km       float64
color     object
price      int64
dtype: object

NUMERICAL SUMMARY
              year             km         price
count   255.000000     255.000000  2.550000e+02
mean   2024.207843   30384.172549  2.247061e+06
std       0.406562   15563.810709  1.517529e+05
min    2024.000000    6001.000000  1.825000e+06
25%    2024.000000   19880.000000  2.160000e+06
50%    2024.000000   29000.000000  2.239750e+06
75%    2024.000000 

# **Understaning the outliers** 


## **year: 53 outliers**

In [4]:
# Calculate fences for year
Q1 = dataset['year'].quantile(0.25)
Q3 = dataset['year'].quantile(0.75)
IQR = Q3 - Q1

lower_fence = Q1 - 1.5 * IQR
upper_fence = Q3 + 1.5 * IQR

# Filter the dataframe to show ONLY the outliers
year_outliers = dataset[(dataset['year'] < lower_fence) | (dataset['year'] > upper_fence)]

# Display the rows
print(year_outliers)

     year       km       color    price
0    2025  69000.0       White  2095000
13   2025  33000.0       Black  2165000
14   2025  39066.0       White  2209750
16   2025   9000.0       White  2395000
19   2025   6001.0  Smoke Gray  2700000
20   2025  17000.0       White  2300000
27   2025  26000.0       White  2185000
29   2025  30000.0       White  2159000
30   2025  35523.0        Gray  2125000
42   2025  25680.0       White  2225000
51   2025  18000.0       White  2250000
56   2025  23000.0       White  2495000
60   2025  15070.0       Black  2251000
74   2025  16500.0       White  2400000
76   2025  41500.0       Black  2180000
78   2025  18600.0       White  2350000
79   2025  15000.0       White  2450000
83   2025  28000.0       Black  2225000
92   2025  39000.0       White  2199900
99   2025  14980.0       Black  2285000
103  2025  19960.0       White  2350000
113  2025  15000.0       White  2450000
115  2025  37000.0       Black  2165000
125  2025  31355.0       Black  2270000


Year outliers (2025) were retained because they represent valid new-model entries rather than data errors. Since the dataset is primarily composed of 2024 models, 2025 values are statistically flagged as outliers, but they are essential for the model to learn the price premium associated with newer vehicles

## **km: 6 outliers**

In [5]:
# Calculate fences for Price
Q1 = dataset['km'].quantile(0.25)
Q3 = dataset['km'].quantile(0.75)
IQR = Q3 - Q1

lower_fence = Q1 - 1.5 * IQR
upper_fence = Q3 + 1.5 * IQR

# Filter the dataframe to show ONLY the outliers
km_outliers = dataset[(dataset['km'] < lower_fence) | (dataset['km'] > upper_fence)]

# Display the rows
print(km_outliers)

     year        km  color    price
0    2025   69000.0  White  2095000
23   2024   73000.0   Blue  2140000
96   2024   75000.0  White  2010000
145  2024  110000.0  White  2040000
210  2024   96000.0  Black  2049999
229  2024   68200.0  White  2100000


In [6]:
print(dataset["km"])

0      69000.0
1      28500.0
2      50000.0
3      42000.0
4      15000.0
        ...   
250    24000.0
251    29000.0
252    19800.0
253    16000.0
254    34000.0
Name: km, Length: 255, dtype: float64


In this dataset, the KM (Mileage) column contains a few rows with exceptionally high values, such as 110,000 km for a 2024 model. This creates three major issues for our Machine Learning model:

1. Statistical Noise: Because the vast majority of our cars have low mileage (around 20,000–30,000 km), these extreme points act as "noise." They pull the mathematical average (mean) away from the typical car, making the data less representative of the standard market.


2. Logic Conflict: A car driven over 100,000 km in a single year is likely a commercial vehicle or a taxi. These cars have significantly more wear-and-tear than private vehicles. Including them confuses the model because their relationship between year and Price is completely different from the rest of the data.


3. Model Skew: When building a Linear Regression model, these distant points can "tilt" the prediction line. The model will try so hard to reach these outliers that it will lose accuracy when predicting prices for "normal" low-mileage cars.

Conclusion

To ensure the model remains accurate and focused on the standard consumer market, these high-mileage outliers must be removed from the dataset.

## **price: 8 outliers**

In [7]:
# Calculate fences for Price
Q1 = dataset['price'].quantile(0.25)
Q3 = dataset['price'].quantile(0.75)
IQR = Q3 - Q1

lower_fence = Q1 - 1.5 * IQR
upper_fence = Q3 + 1.5 * IQR

# Filter the dataframe to show ONLY the outliers
price_outliers = dataset[(dataset['price'] < lower_fence) | (dataset['price'] > upper_fence)]

# Display the rows
print(price_outliers)

     year       km       color    price
17   2024  26000.0       White  1835000
19   2025   6001.0  Smoke Gray  2700000
22   2024  47500.0       White  1880000
36   2024  17000.0       White  1899750
170  2025  30000.0       White  3100000
199  2025  21000.0       Black  2969000
213  2025  16000.0       White  2850000
221  2024  32000.0       Black  1825000


In [8]:
print(dataset["price"])

0      2095000
1      2280000
2      2090000
3      2180000
4      2300000
        ...   
250    2390000
251    2185000
252    2305000
253    2250000
254    2310000
Name: price, Length: 255, dtype: int64


Why Price Outliers are Kept

The high Price values (2.7M–3.1M) represent valid market premiums for new 2025 models. Unlike errors, these are "high-quality" outliers that show the maximum value a car can reach. Removing them would make your model under-predict the value of newer cars.

# **Cleaning the data** 

## **Remove km outliers (6) & remove the duplication (4)**

6 + 4 = 10 Rows

255 - 10 = 245 Rows

shape = 245 * 4 

In [9]:
# Remove rows that exist in km_outliers from the original dataset
dataset_clean = dataset.drop(km_outliers.index).copy()

# Step 2: Remove duplicates FROM the cleaned dataset
dataset_clean = dataset_clean.drop_duplicates().copy()

# Final result
print(dataset_clean)

     year       km  color    price
1    2024  28500.0  Black  2280000
2    2024  50000.0  White  2090000
3    2024  42000.0  White  2180000
4    2024  15000.0  White  2300000
5    2024  36326.0  White  2155000
..    ...      ...    ...      ...
250  2025  24000.0  Black  2390000
251  2024  29000.0  White  2185000
252  2025  19800.0  White  2305000
253  2024  16000.0  White  2250000
254  2024  34000.0  White  2310000

[245 rows x 4 columns]


In [10]:
print(f"Rows: {dataset_clean.shape[0]}, Columns: {dataset_clean.shape[1]}")

Rows: 245, Columns: 4


# **Select Independent Variable (X) and Dependent Variable (y)**

📌 Note: The concepts of Independent and Dependent variables were discussed earlier.

If you would like to review or gain a clearer understanding, please refer to this section: [Go to Independent variable VS Dependent variable Section](https://github.com/Hazem1695/ml-concept-briefs)

In [11]:
# Separate independent variable(features (X)) by selecting all rows and all columns except the last one
X = dataset_clean.iloc[:, :-1].values
# Separate the dependent variable (Target(y)) by selecting the last column
y = dataset_clean.iloc[:, -1].values

In [12]:
print(X)

[[2024 28500.0 'Black']
 [2024 50000.0 'White']
 [2024 42000.0 'White']
 [2024 15000.0 'White']
 [2024 36326.0 'White']
 [2024 16500.0 'Blue']
 [2024 13500.0 'White']
 [2024 34000.0 'Silver Gray']
 [2024 32000.0 'White']
 [2024 14550.0 'White']
 [2024 31000.0 'Burgundy']
 [2024 15000.0 'White']
 [2025 33000.0 'Black']
 [2025 39066.0 'White']
 [2024 44000.0 'Blue']
 [2025 9000.0 'White']
 [2024 26000.0 'White']
 [2024 55000.0 'White']
 [2025 6001.0 'Smoke Gray']
 [2025 17000.0 'White']
 [2024 41500.0 'White']
 [2024 47500.0 'White']
 [2024 29700.0 'White']
 [2024 45000.0 'White']
 [2024 62000.0 'White']
 [2025 26000.0 'White']
 [2024 13500.0 'Black']
 [2025 30000.0 'White']
 [2025 35523.0 'Gray']
 [2024 12800.0 'White']
 [2024 32934.0 'Black']
 [2024 53000.0 'White']
 [2024 30000.0 'White']
 [2024 28500.0 'White']
 [2024 17000.0 'White']
 [2024 26000.0 'White']
 [2024 30000.0 'Black']
 [2024 38000.0 'White']
 [2024 20500.0 'White']
 [2024 20000.0 'White']
 [2025 25680.0 'White']
 [2024 

In [13]:
print(y)

[2280000 2090000 2180000 2300000 2155000 2300000 2280000 2350000 2075000
 2290000 2325000 2239750 2165000 2209750 2225000 2395000 1835000 2125000
 2700000 2300000 2300000 1880000 2280000 2120000 1995000 2185000 2390000
 2159000 2125000 2300000 2225000 2099000 2400000 2199000 1899750 2225000
 2280000 1950000 2250000 2435500 2225000 2150000 2275000 2210000 1990000
 2189000 2125000 2190000 2350000 2250000 2490000 2325000 2435000 2110000
 2495000 2249990 2098500 2220000 2251000 2195000 2250000 2150000 2110000
 2185000 2050000 2336000 2320000 2300000 2329000 2200000 2185000 2185000
 2400000 2275000 2180000 2390000 2350000 2450000 2165000 2189750 2125000
 2225000 2230000 2085000 2160000 2065000 2200000 2135000 2280000 2490000
 2199900 2240000 2185000 2270000 2150000 2455000 2285000 2035000 2520000
 2339000 2350000 2145000 2298000 2195000 2350000 2215000 2150000 2100000
 2125000 2239000 2220000 2165000 2290000 2249900 2170000 2250000 2180000
 2375000 2200000 2300000 2220000 2270000 2250000 23

# **Encoding categorical data**

## Encoding the Indpendent Variable Using One Hot Encoding

📌 Note: I explained One-Hot Encoding earlier.

If you missed it or want a deeper understanding of how it works, you can find it here: [Go to One-Hot Encoding Section](https://github.com/Hazem1695/ml-concept-briefs)

In [14]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
# One-hot encode the 'color' column (index 2) while keeping other features
# Initialize ColumnTransformer to handle categorical data
# [2] targets the 'color' column (index 2)
ct = ColumnTransformer(transformers = [('encoder', OneHotEncoder(sparse_output=False),[2])], remainder = 'passthrough')
# Execute transformation and convert to NumPy ndarray for model training
X = np.array(ct.fit_transform(X))

In [15]:
print(X)

[[1.0 0.0 0.0 ... 0.0 2024 28500.0]
 [0.0 0.0 0.0 ... 1.0 2024 50000.0]
 [0.0 0.0 0.0 ... 1.0 2024 42000.0]
 ...
 [0.0 0.0 0.0 ... 1.0 2025 19800.0]
 [0.0 0.0 0.0 ... 1.0 2024 16000.0]
 [0.0 0.0 0.0 ... 1.0 2024 34000.0]]


# **Splitting the dataset into the Train set and Test set**

In [16]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train , y_test = train_test_split(X, y , test_size = 0.2, random_state = 0)

In [17]:
print(X_train)

[[0.0 0.0 0.0 ... 1.0 2025 9000.0]
 [0.0 0.0 1.0 ... 0.0 2024 41250.0]
 [0.0 0.0 0.0 ... 0.0 2024 34000.0]
 ...
 [0.0 0.0 0.0 ... 1.0 2024 16000.0]
 [0.0 0.0 0.0 ... 1.0 2024 36000.0]
 [0.0 0.0 0.0 ... 1.0 2025 12000.0]]


In [18]:
print(X_test)

[[0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 25000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 22700.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 10124.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 33000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 66000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 41300.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 7650.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2024 6001.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 37000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 44000.0]
 [0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 2024 31000.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2025 19200.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 6001.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2025 19200.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2024 52531.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2025 41200.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 29000.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2024 25000.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2025 41500.0]
 [0.0 1.0 0.0 0

In [19]:
print(y_train)

[2395000 2279000 2350000 2100000 2225000 2195990 2400000 2150000 2250000
 2085000 2199000 2029000 2250000 2035000 2310000 2199900 2155000 2310000
 2850000 2270000 1835000 2125000 2250000 2350000 2155000 2220000 2150000
 2200000 2435000 2320000 2390000 2130000 2110000 2249999 2225000 2209750
 2300000 2145000 1995000 2225000 2305000 2250000 2098500 2250000 2180000
 2300000 2370000 2150000 2495000 2185000 2125000 2325000 2520000 2300000
 2099000 2200000 2325000 2285750 2250000 2209000 2150000 2225000 2159000
 2490000 2235000 2300000 2220000 2125000 2090000 2390000 2049999 2300000
 2110000 2079000 2320000 2180000 2195000 2120000 2250000 2210000 2300000
 2160000 2235000 2290000 2280000 2520000 2257500 2270000 2325000 2325000
 2249900 2280000 2150000 2455000 2200000 2250000 2350000 2065000 2305000
 2349000 2120000 2150000 2385000 2165000 2350000 2295000 2115000 2250000
 2280000 2329000 2160000 2270000 2239000 2390000 1825000 2120000 2350000
 2240000 2100000 2239750 2300000 2298000 2225000 22

In [20]:
print(y_test)

[2185000 2250000 2275000 2249990 2089000 2199000 2350000 2425000 2100000
 2050000 2320000 2350000 2490000 2365000 2125000 2400000 2265555 2340000
 2180000 2350000 2165000 2189000 2185000 2215000 2199000 2240000 2199000
 2075000 2500000 2285000 1990000 2130000 2300000 2280000 2165000 1950000
 2490000 2269900 2195000 2240000 2390000 2200000 2160000 2270000 2185000
 2240000 2125000 2700000 2239000]


# **Building Multiple Linear Regression Model**

## Training the Multiple Linear Regression model on the Training set

In [21]:
from sklearn.linear_model import LinearRegression
"""
OOP
regressor is object
LinearRegression() is class
LinearRegression() no need to enter any parameters
"""
regressor = LinearRegression()
# train our model
regressor.fit(X_train, y_train)

LinearRegression()

# **Preticting the Test set results**

In [22]:
"""
Note:
y_test: contains the real profits
y_pred: contains the predicted profits
"""
# test our model
y_pred = regressor.predict(X_test)
"""
Note:
we did it only to compare betwen predict profits with real profits
to see how our model work (correct or wrong)
"""
np.set_printoptions(precision=2)
print(np.concatenate((y_pred.reshape(len(y_pred), 1), y_test.reshape(len(y_test), 1)), axis = 1))

[[2232496.22 2185000.  ]
 [2242772.07 2250000.  ]
 [2298958.63 2275000.  ]
 [2196754.13 2249990.  ]
 [2049318.03 2089000.  ]
 [2159671.72 2199000.  ]
 [2310011.87 2350000.  ]
 [2358641.46 2425000.  ]
 [2178883.09 2100000.  ]
 [2147608.77 2050000.  ]
 [2290374.08 2320000.  ]
 [2398201.09 2350000.  ]
 [2317379.2  2490000.  ]
 [2356938.83 2365000.  ]
 [2150756.55 2125000.  ]
 [2258648.1  2400000.  ]
 [2214625.18 2265555.  ]
 [2273758.47 2340000.  ]
 [2298570.02 2180000.  ]
 [2316908.31 2350000.  ]
 [2318674.95 2165000.  ]
 [2241303.03 2189000.  ]
 [2233836.55 2185000.  ]
 [2295283.73 2215000.  ]
 [2233548.63 2199000.  ]
 [2211050.97 2240000.  ]
 [2257068.9  2199000.  ]
 [2201221.89 2075000.  ]
 [2317374.74 2500000.  ]
 [2417055.04 2285000.  ]
 [2174415.33 1990000.  ]
 [2223560.7  2130000.  ]
 [2337013.23 2300000.  ]
 [2211497.74 2280000.  ]
 [2336545.99 2165000.  ]
 [2174415.33 1950000.  ]
 [2299512.63 2490000.  ]
 [2169947.57 2269900.  ]
 [2098463.4  2195000.  ]
 [2304219.26 2240000.  ]


# **Single Prediction**

In [ ]:
X_new = [[2024, 30000,'White']]

X_new_transformed = ct.transform(X_new)

prediction = regressor.predict(X_new_transformed)

print(prediction)

[2210157.41]


# **Evaluating the Model Performance**

In [24]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred)

0.4203676854917229

# **linear regression equation with coefficients**

In [25]:
print(regressor.coef_)
print(regressor.intercept_)

[-2.00e+04  5.33e+03  2.35e+04 -9.38e+04  3.85e+04  6.24e+04  4.53e+04
  7.28e-12 -6.12e+04  9.85e+04 -4.47e+00]
-197018509.885121


## Multiple Linear Regression Equation


$\begin{aligned} \hat{y} = &\ -197018509.89 \\ &- 20000 \cdot x_1 \\ &+ 5330 \cdot x_2 \\ &+ 23500 \cdot x_3 \\ &- 93800 \cdot x_4 \\ &+ 38500 \cdot x_5 \\ &+ 62400 \cdot x_6 \\ &+ 45300 \cdot x_7 \\ &+ 0 \cdot x_8 \\ &- 61200 \cdot x_9 \\ &+ 98500 \cdot x_{10} \\ &- 4.47 \cdot x_{11} \end{aligned}$ 



**Where:**

- $(x_1, x_2, ..., x_9)$: One-hot encoded color features  
- $(x_{10})$: Year  
- $(x_{11})$: KM (mileage)

# **Conclusion**

Applying feature scaling did not significantly improve the model performance. This aligns with the theoretical understanding that Multiple Linear Regression is not sensitive to feature scaling, as the model coefficients adjust to the scale of the input features. However, feature scaling will be important in later models such as Support Vector Regression.
